### ISSUE TITLE
Puma misses some fire-and-forget requests

### ISSUE THREAD

**Issue by ilyazub** [puma/puma#2624]
**Describe the bug**
`puma` misses fire-and-forget requests created with `TCPSocket.open`. `sleep(0.001)` before closing the socket connection helps `puma` to catch these requests. Incorrect fire-and-forget requests without `sleep` are received and displayed in logs unlike correct requests.

**Puma config:**

&mdash;

**To Reproduce**

1. Create a `hello.ru` file:

```ruby
run lambda { |env| [200, {"Content-Type" => "text/plain"}, ["Hello World"]] }
```

2. Run it with:

```bash
bundle exec puma -v --debug hello.ru
```

3. Send 10 HTTP request without waiting for the response:

```bash
ruby -r socket -e '10.times { TCPSocket.open("0.0.0.0", 9292) { |c| c.puts "GET / HTTP/1.1\r\n\r\n"; sleep 0.00009999 } }'
```

4. Check in the `puma` logs that four of ten requests being received:

```bash
127.0.0.1 - - [07/May/2021:17:27:59 +0300] "GET / HTTP/1.1" 200 - 0.0001
127.0.0.1 - - [07/May/2021:17:27:59 +0300] "GET / HTTP/1.1" 200 - 0.0002
127.0.0.1 - - [07/May/2021:17:27:59 +0300] "GET / HTTP/1.1" 200 - 0.0001
127.0.0.1 - - [07/May/2021:17:27:59 +0300] "GET / HTTP/1.1" 200 - 0.0001
```

The same problem happens without block-form of sending the message:

```bash
ruby -r socket -e '10.times { c = TCPSocket.open("0.0.0.0", 9292); c.puts "GET / HTTP/1.1\r\n\r\n"; sleep 0.00009999; c.close }'
```

Replacing `puts` with `write` leads to the same behavior.

```bash
ruby -r socket -e '10.times { c = TCPSocket.open("0.0.0.0", 9292); c.write "GET / HTTP/1.1\r\n\r\n"; sleep 0.00009999; c.close }'
```

**Expected behavior**
`puma` not misses fire-and-forget requests.

**Desktop (please complete the following information):**
 - OS: Fedora release 33 (Thirty Three)
 - Puma Version: 5.2.2

Same problem with Puma v4.3.6.

**Comment by nateberkopec** [comment]
Is the problem, from your perspective, that Puma doesn't _log_ these connections? Because as Greg said, we don't response to sockets that are closed.

**Comment by ilyazub** [comment]
> This is an easy fix, but since 'fire-and-forget' requests could also be used to swamp a server, we might need a config option like `allow_fire_and_forget`.
> 
> Puma won't send the request to the app if the request socket is closed, so it's a matter of timing.
> 
> Adding the above option would move that check after the request is processed by the app.
> 
> Not sure what others think...

Yes, it could swamp a server. We successfully use such requests for several years in production and some requests were missed only on my machine.

In what file _that check_ is placed? I'll patch it locally and test fire-and-forget requests again.

**Comment by ilyazub** [comment]
> Is the problem, from your perspective, that Puma doesn't _log_ these connections? Because as Greg said, we don't respond to sockets that are closed.

The problem is that Puma doesn't handle these connections. Logging is just an illustration of unexpected (for me) behavior.

**Comment by calvinxiao** [comment]
> In what file that check is placed? I'll patch it locally and test fire-and-forget requests again.

@ilyazub please check here https://github.com/puma/puma/blob/6dfb8bc2ba1175198f5982cc8092bcb7f021fe22/lib/puma/request.rb#L35

**Comment by ilyazub** [comment]
> > Puma won't send the request to the app if the request socket is closed, so it's a matter of timing.
>
> Some requests were missed only on my machine — other developers on macOS and Windows haven't that problem.

All fire-and-forget requests are handled on a dummy server on Ubuntu 20.04. Just tried to reproduce that problem non-locally.

**Comment by MSP-Greg** [comment]
@ilyazub

What @calvinxiao said (thanks).  You still need that line of code, it just needs to moved somewhere after line 77.

This feature is a good example of something some users want, but would rarely be used by standard 'Rails, behind nginx' setups...

**Comment by MSP-Greg** [comment]
> If you agree with the addition of this feature

I do, maybe allow some time for other maintainers to respond.

>`return false if closed_socket?(io) && !@options[:allow_fire_and_forget]`

Maybe check `!@options[:allow_fire_and_forget]` first?

**Comment by nateberkopec** [comment]
>I do, maybe allow some time for other maintainers to respond.

I don't get it. I can understand wanting it logged, but why would you want to respond to a socket that's cclosed?

> Some requests were missed only on my machine — other developers on macOS and Windows haven't that problem

Can you expand on this? I don't understand the problem or the use case.

**Comment by MSP-Greg** [comment]
> but why would you want to respond to a socket that's closed?
 
You did say you googled it, but to clarify for others that don't...

The issue is not about responding to a closed socket, but passing the request to the app, regardless of whether the socket is closed after the request is sent/received.

**Comment by hartator** [comment]
> The issue is not about responding to a closed socket, but passing the request to the app, regardless of whether the socket is closed after the request is sent/received.

Yes, exactly. To be clear, this is **the current observed behavior** of Puma. To allow fire and forget HTTP requests. It works fine on all our OSX and Win machines as well as our Ubuntu production servers. We use it to easily load balance internal background tasks.

Just Ilya's machine behave in odd way where random requests get dropped. It might be a race condition somewhere? Like maybe his Linux OS configs make some part of Puma run slow before `lib/puma/request.rb:35` is reached?

I think it might be more valuable for us and Puma to find out the root cause of particular slowness instead of adding a flag. As setting the `allow_fire_and_forget` flag to false will be an indirection and give a misguided sense of security as it won't disallow these requests anyway.

**Comment by MSP-Greg** [comment]
@hartator

> Yes, exactly. To be clear, this is the current behavior of Puma. To allow fire and forget HTTP requests.

I just verified that with WSL2/Ubuntu 20.04, no 'fire and forget' requests are processed by the app.  That's using the single line in the first post with no sleep.

I had the same result yesterday, but after your message I wanted to double check.

> I think it might be more valuable for us and Puma to find out the root cause of particular slowness

I don't think it's a Puma issue.  Sockets take finite amounts of time.  They can be interrupted or slowed down, especially on systems with multiple VMs on a single physical system.

I've been working on a test suite update, with some tests creating 2k http requests.  Several of those tests always pass locally with no read errors, but there are often read errors when run on GitHub Actions.

**Comment by MSP-Greg** [comment]
To clarify a few more things:

1. This behavior on Ubuntu is affected by the location of the `closed_socket?(io)` call.  That method only works on Ubuntu, as both macOS & Windows don't support:
    ```ruby
    tcp_socket.getsockopt(Socket::ANY_CONST, Socket::TCP_INFO)
    ```
    Note that the OS's may support something similar, but Ruby does not.


2. On macOS & Windows Ruby, I believe the only way for a server to determine if a client has closed a socket is to read or write to it.  There is no way to check for closure by any type of inspection.

**Comment by ilyazub** [comment]
> 1. This behavior on Ubuntu is affected by the location of the `closed_socket?(io)` call.

After moving `closed_socket?(io)` below `@app.call(env)`, 'fire-and-forget' requests without `sleep` are handled.

> That method only works on Ubuntu, as both macOS & Windows don't support:

The detection of a closed socket is supported in Ruby on Ubuntu 18-20 and Fedora 32-33.

```bash
$ ruby -r socket -e 'puts Socket.const_defined?(:TCP_CORK) && Socket.const_defined?(:IPPROTO_TCP)'
true
```


Anyway, I'm closing this issue because we'll move our internal background tasks to background jobs and there's no need to make Puma more complex.

----

Thank you all for your responsiveness and sorry for distracting you in vain.
